sklearn/cluster/_kmeans.py

```diff
@@ -59,15 +59,16 @@
 # Initialization heuristic
 
 
-@validate_params(
-    {
-        "X": ["array-like", "sparse matrix"],
-        "n_clusters": [Interval(Integral, 1, None, closed="left")],
-        "x_squared_norms": ["array-like", None],
-        "random_state": ["random_state"],
-        "n_local_trials": [Interval(Integral, 1, None, closed="left"), None],
-    }
-)
+_parameter_constraints_of_kmeans_plusplus = {
+    "X": ["array-like", "sparse matrix"],
+    "n_clusters": [Interval(Integral, 1, None, closed="left")],
+    "x_squared_norms": ["array-like", None],
+    "random_state": ["random_state"],
+    "n_local_trials": [Interval(Integral, 1, None, closed="left"), None],
+}
+
+
+@validate_params(_parameter_constraints_of_kmeans_plusplus)
```

Comment:
Could this directly be inlined?

```suggestion
@validate_params({
    "X": ["array-like", "sparse matrix"],
    "n_clusters": [Interval(Integral, 1, None, closed="left")],
    "x_squared_norms": ["array-like", None],
    "random_state": ["random_state"],
    "n_local_trials": [Interval(Integral, 1, None, closed="left"), None],
})
```

Reply:
No, that's why I extracted it. To be able to test it, we need to access the dict of constraints in the test. It's theoretically possible to recover the parameters of a decorator programmatically but it requires inspecting the code of the function and reconstruct it. It's not easy and would likely be unreadable, bring many bugs, and be hard to maintain. I think having the constraint right next to the function is an acceptable compromise.